app/lobby/server/game_proposer.js

```diff
@@ -5,11 +5,27 @@ GameProposer = class GameProposer {
     this.exclusions = exclusions
   }
 
-  propose() {
-    this.select_cards()
-    this.proposal_id = this.create_proposal()
-    this.update_players()
-    this.set_proposal_timeout()
+  propose(kingdom_id) {
```

Comment:
If we capture the `kingdom_id` in the constructor like the other values are, then that becomes:

```
constructor(player_ids, exclusions, kingdom_id) {
  this.players = this.find_players(player_ids)
  this.exclusions = exclusions
  this.kingdom_id = kingdom_id
}
```

With the validation pushed to the client side and the `kingdom_id` fed into the constructor, the `propose` function can go back to it's original form, and still handle both cases, as:

```
propose(kingdom_id) {
  this.select_cards()
  this.proposal_id = this.create_proposal()
  this.update_players()
  this.set_proposal_timeout()
}
```

Then we only need a single `select_cards` function, which can be:

```
card_list = new CardList(this.exclusions)
if (this.kingdom_id === '') {
  this.cards = card_list.pull_set()
} else {
  this.cards = card_list.pull_from_history(this.kingdom_id)
}
```